[py] `DecRef` all owned references to `PyObject`s

olivielpeau opened the pull request:
### What does this PR do?

* `DecRef`s all owned references to `PyObject`s in py's `Check.Run()` method
* Add a test on `AgentCheck.warning` (not directly related to the 1st change)

### Motivation

Since we added `warning` to `AgentCheck` in #396 there's a pretty significant memory leak which (I think) comes from not properly decreasing the reference count on the result of `c.Instance.CallMethod("get_warnings", emptyTuple)` (a list) that is owned by the caller.

### Additional Notes

More information on reference count and ownership of references when using the python C API: https://docs.python.org/2/c-api/intro.html#objects-types-and-reference-counts

### PR Updated

I went through all our python C API calls and added `DecRef`/`IncRef` wherever needed, except for 2 tricky cases that I've documented.

I've also removed `getPythonError` (copy/paste of `stickyLock.getPythonError` which should be preferred).

olivielpeau committed:
[py] `DecRef` all owned references to `PyObject`s

masci reviewed (approved):
Good catch, I'm responsible for leaving most of that garbage around 👼

olivielpeau committed:
[py] Systematically fix (almost) all ref leaks

Except one reference leak in PythonCheckLoader.Load() and the fact
that we don't `DecRef` the fields of `PythonCheck` when we discard one,
this commit should fix all python reference leaks.

masci reviewed (approved):
Thanks for this HUGE work!

olivielpeau committed:
[py] Set finalizer on PythonCheck to `DecRef` on owned PyObject refs

olivielpeau committed:
Merge branch 'master' into olivielpeau/warnings-mem-leak
